src/normalizer/mod.rs

```diff
@@ -78,17 +72,51 @@ pub trait Normalizer: Sync + Send {
     /// Normalize the provided [`Token`].
     /// Options can be set using the provided [`NormalizerOption`].
     ///
-    /// A Normalizer can return several `Token`s.
-    fn normalize<'o>(
-        &self,
-        token: Token<'o>,
-        options: NormalizerOption,
-    ) -> Box<dyn Iterator<Item = Token<'o>> + 'o>;
+    fn normalize<'o>(&self, mut token: Token<'o>, options: NormalizerOption) -> Token<'o> {
+        if options.create_char_map {
+            match token.char_map.take() {
+                Some(mut char_map) => {
+                    let mut lemma = String::new();
+                    let mut tail = token.lemma.as_ref();
+                    for (_, normalized_len) in char_map.iter_mut() {
+                        let (head, t) = tail.split_at(*normalized_len as usize);
+                        tail = t;
+                        let normalized = self.normalize_str(head);
+                        *normalized_len = normalized.len() as u8;
+                        lemma.push_str(normalized.as_ref());
+                    }
```

Comment:
Trying to understand what is happening here, for myself:
1. We look at every character of the original lemma.
2. For every character, we normalise it using `normalise_str`
3. Then we change the length of this character in the `char_map`, based on the length of the normalised char
4. And we push the normalised char to the new lemma

That looks good :-)

I wonder, however, what a "character" from the `char_map` is. Is it always a single unicode code point (a `char` in Rust) or can it be a grapheme cluster? 

In either case, I think there is a risk of misunderstanding about the kind of inputs that `normalise_str` accepts. I initially expected it to accept whole strings such as `lion`, but it seems like it is mostly given single characters unless `options.create_char_map` is false.

If a character in the `char_map` is guaranteed to be a `char`, then maybe it would be worth having a `normalize_char` trait function? It could clarify that a normaliser shouldn't expect to have access to a whole lemma at once, and could potentially improve performance in some cases.

Reply:
> I wonder, however, what a "character" from the char_map is. Is it always a single unicode code point (a char in Rust) or can it be a grapheme cluster?

It's always a char initially. Because in Meilisearch we work with normalized words but we want to highlight original words, the goal of the char map is to retrieve the original character's "position" in order to highlight the original word properly based on the number of characters that should be highlighted in the normalized version.

> In either case, I think there is a risk of misunderstanding about the kind of inputs that normalise_str accepts. I initially expected it to accept whole strings such as lion, but it seems like it is mostly given single characters unless options.create_char_map is false.

It's not always the case, the method normalize is overridden in the japanese and the lowercase normalizers skip the char map creation and call `normalize_str` over the whole string.

> If a character in the char_map is guaranteed to be a char, then maybe it would be worth having a normalize_char trait function? It could clarify that a normaliser shouldn't expect to have access to a whole lemma at once, and could potentially improve performance in some cases.

Yes but no, "yes" for the first normalizer that creates a char map, then "no" because the next normalizers may receive more than one character corresponding to the original character.
For instance, `deunicode` transforms 🦄 into "unicorn" and then the next normalizer will receive the whole "unicorn" string corresponding to the original character 🦄. It's not the best example I found sorry 😬